provisioning/miniprov/hare_mp/utils.py

```diff
@@ -193,21 +193,20 @@ def stop_hare(self):
     def is_hare_stopping(self) -> bool:
         return self.hare_stop
 
-    def get_io_nodes(self) -> List[str]:
+    def get_nodes_from_services(self, service_type: str, with_hostname=False,
```

Comment:
What does this method actually return?

As far as I can see, it returns a list of strings. The list of strings contains service IDs (or names?) AND potentially - machine_ids. Is my understanding correct?

If yes, I wonder how the end developers will use the function? I'm afraid it will be very hard to distinguish which list element means what.

Reply:
This function will return 3 types of list. You can understand it by it's calling mechanism and it's output like below:
```
utils.get_nodes_from_services(Const.SERVICE_MOTR_IO.value, with_machine_id=True)
output:
		['aaa120a9e051d103c164f605615c32a4', 'bbb340f79047df9bb52fa460615c32a5', 'ccc8700fe6797ed532e311b0615c32a7']

utils.get_nodes_from_services(Const.SERVICE_MOTR_IO.value, with_hostname=True)
output:
		['data-node1', 'data-node2', 'data-node3']
		
utils.get_nodes_from_services(Const.SERVICE_MOTR_IO.value)
output:
		['node>aaa120a9e051d103c164f605615c32a4>components[1]>services[0]', 'node>bbb340f79047df9bb52fa460615c32a5>components[1]>services[0]', 'node>ccc8700fe6797ed532e311b0615c32a7>components[1]>services[0]']

```

Reply:
> This function will return 3 types of list

Why don't you introduce 3 different functions (with corresponding names) then?

If the function is named "get_nodes_from_services", then I would expect that it returns nodes only. In your case it can return almost anything 😁 

In UX design, there is a well known "minimal astonishment" principle. For instance, if the button is labelled with "Door bell" inscription, it should not launch a rocket to the Moon. It should only cause the ring to bell. I think that the same principle should be considered in software design to make the life of the teammates comfortable.

Reply:
Since we are using same small piece of code in multiple places. It's more over look like a redundancy of code.
So regarding the same, I had a discussion with @mssawant on the same topic and we had agreed upon a solution where we can optimize such code with single function (snippet) which having similar business logic. 
     I have also provided enough flexibility to this function to extract list based on requirement .

1. If you are looking for the list of nodes with its "hostname" w.r.t service type 
    then, it will return a list like :   ['data-node1', 'data-node2', 'data-node3']
2. if you want a list of nodes with its "machine_id" w.r.t service type
    then, it will return : ['aaa120a9e051d103c164f605615c32a4', 'bbb340f79047df9bb52fa460615c32a5', 'ccc8700fe6797ed532e311b0615c32a7']
 3. If you doesn't specify either "hostname" or "machine_id" 
     then, it simply return a default list like :
     ['node>aaa120a9e051d103c164f605615c32a4>components[1]>services[0]', 'node>bbb340f79047df9bb52fa460615c32a5>components[1]>services[0]', 'node>ccc8700fe6797ed532e311b0615c32a7>components[1]>services[0]']
     
Here, we could work with the naming convention of the function or its arguments if required.

Reply:
@rahul27kumar As I pointed out, the name is misleading (there is a combination of flags that can change the behavior of the function significantly, so that the name stops being meaningful). And even the resulting type wouldn't help the developers to uncover a mistake while using your function (reason: when everything is a string, there is no way to catch the problem using static analysis tools like mypy).

> It's more over look like a redundancy of code.

If redundant code is less error-prone than, so to say, optimized, then I definitely vote for a redundant version.

My point is that humans are not robots. We make mistakes. It is nice when there is a way to identify a mistake either at build time or during the code review. Your approach makes the code little bit more cryptic so some mistakes can be hidden during code review and will bump into issues in the field only - which is definitely more expensive to debug and fix.


> Here, we could work with the naming convention of the function or its arguments if required.

If you have any suggestions how to address the points I mentioned a few lines above, let's discuss. 

cc @vaibhavparatwar, @SunilSavanurSeagate